### ISSUE TITLE
Headers from S3 object?

### ISSUE THREAD

**Issue by GeoffreyPlitt** [izaakschroeder/s3-streams#13]
I'm trying to stream the S3 object to the browser as a response, with correct Content-Type header.

I've got an S3 type with the Content-Type header, and I do:
```javascript
  const stream = S3S.ReadStream(new S3(), {
    Bucket: config.aws_upload_bucket,
    Key: key
  })
  stream.pipe(res)
```
There doesn't seem to be any Content-Type, and I couldn't even find any headers, on the stream object.

**Comment by izaakschroeder** [comment]
Oh this is kind of a lovely idea!

How about something like this?

```javascript
// You also have access to `Bucket` and `Key`
stream.once('open', function(file) {
  res.setHeaders({
    'Content-Type': file.ContentType,
    'Content-Length': file.ContentLength
  });
});
```

**Comment by izaakschroeder** [comment]
I suppose it might be an idea for S3 streams to just hook `pipe` and "do the right thing" automatically. Will have to give it some thought. I think the above will work for now though.

**Comment by izaakschroeder** [comment]
The native AWS SDK also has reasonable support for streaming reads (and they offer a similar event, `httpHeaders` that includes the raw headers to pluck data out of) if you'd rather use that.

**Comment by GeoffreyPlitt** [comment]
Cool. Yeah I wanted to use `pipe()` instead of events. Letting pipe set the headers sounds like amazing magic to me; I suppose it might be better to separate them though as a more flexible solution.

The only problem with the API you announced above is that the headers must be set before the body starts piping, so I'd put the `stream.pipe()` call within the `stream.once()` callback, not outside it.

**Comment by izaakschroeder** [comment]
I'm giving smart piping a shot. I still have to finish writing tests and whatnot, but feel free to give the branch a whirl and see if it meets your needs.

**Comment by izaakschroeder** [comment]
Alright well tests are done. If you confirm the branch does what you want I'll merge it in and publish the updated version on npm shortly after.

**Comment by GeoffreyPlitt** [comment]
Left comments on the PR. Huge props for replying so quick, love it.

**Comment by izaakschroeder** [comment]
Smart piping available in `s3-streams@0.3.0`. :beers:

**Comment by GeoffreyPlitt** [comment]
+1

On Tue, Jun 16, 2015 at 11:10 PM, Izaak Schroeder <notifications@github.com>
wrote:

> Smart piping available in s3-streams@0.3.0. [image: :beers:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/izaakschroeder/s3-streams/issues/13#issuecomment-112666401>
> .
>



-- 
http://www.geoffplitt.com
http://facebook.com/geoffrey.plitt
https://twitter.com/GeoffreyPlitt
773.339.0915
